src/crypto/ocb_openssl.cc

```diff
@@ -1,1206 +1,162 @@
-        /* Process remaining associated data, compute its tag contribution */
-        remaining = ((unsigned)ad_len) % (BPI*16);
-        if (remaining) {
-			k=0;
-			#if (BPI == 8)
-			if (remaining >= 64) {
-				tmp.bl = xor_block(ad_offset, ctx->L[0]);
-				ta[0] = xor_block(tmp.bl, adp[0]);
-				tmp.bl = xor_block(tmp.bl, ctx->L[1]);
-				ta[1] = xor_block(tmp.bl, adp[1]);
-				ad_offset = xor_block(ad_offset, ctx->L[1]);
-				ta[2] = xor_block(ad_offset, adp[2]);
-				ad_offset = xor_block(ad_offset, ctx->L[2]);
-				ta[3] = xor_block(ad_offset, adp[3]);
-				remaining -= 64;
-				k=4;
-			}
-			#endif
-			if (remaining >= 32) {
-				ad_offset = xor_block(ad_offset, ctx->L[0]);
-				ta[k] = xor_block(ad_offset, adp[k]);
-				ad_offset = xor_block(ad_offset, getL(ctx, ntz(k+2)));
-				ta[k+1] = xor_block(ad_offset, adp[k+1]);
-				remaining -= 32;
-				k+=2;
-			}
-			if (remaining >= 16) {
-				ad_offset = xor_block(ad_offset, ctx->L[0]);
-				ta[k] = xor_block(ad_offset, adp[k]);
-				remaining = remaining - 16;
-				++k;
-			}
-			if (remaining) {
-				ad_offset = xor_block(ad_offset,ctx->Lstar);
-				tmp.bl = zero_block();
-				memcpy(tmp.u8, adp+k, remaining);
-				tmp.u8[remaining] = (unsigned char)0x80u;
-				ta[k] = xor_block(ad_offset, tmp.bl);
-				++k;
-			}
-			AES_ecb_encrypt_blks(ta,k,&ctx->encrypt_key);
-			switch (k) {
-				#if (BPI == 8)
-				case 8: ad_checksum = xor_block(ad_checksum, ta[7]);
-					/* fallthrough */
-				case 7: ad_checksum = xor_block(ad_checksum, ta[6]);
-					/* fallthrough */
-				case 6: ad_checksum = xor_block(ad_checksum, ta[5]);
-					/* fallthrough */
-				case 5: ad_checksum = xor_block(ad_checksum, ta[4]);
-					/* fallthrough */
-				#endif
-				case 4: ad_checksum = xor_block(ad_checksum, ta[3]);
-					/* fallthrough */
-				case 3: ad_checksum = xor_block(ad_checksum, ta[2]);
-					/* fallthrough */
-				case 2: ad_checksum = xor_block(ad_checksum, ta[1]);
-					/* fallthrough */
-				case 1: ad_checksum = xor_block(ad_checksum, ta[0]);
-			}
-			ctx->ad_checksum = ad_checksum;
-		}
-	}
+int ae_decrypt(ae_ctx *ctx, const void *nonce_ptr, const void *ct_ptr,
+	       int ct_len, const void *ad_ptr, int ad_len, void *pt_ptr,
+	       const void *tag, int final) {
+  int len = 0;
+  int plaintext_len = 0;
+  const unsigned char *nonce = (const unsigned char *)nonce_ptr;
+  const unsigned char *ct = (const unsigned char *)ct_ptr;
+  const unsigned char* ad = (const unsigned char *)ad_ptr;
+  unsigned char* pt = (unsigned char *)pt_ptr;
+  if (tag == NULL) {
+    ct_len -= ctx->tag_len;
+  }
+  // Set the key and nonce (IV) to the existing context if a nonce was
+  // provided. It will be omitted if `final` is set to AE_PENDING and
+  // "streaming" operation is in use.
+  if (nonce != NULL &&
+      EVP_DecryptInit_ex(ctx->ocb_ctx, NULL, NULL, ctx->key, nonce) != 1) {
+    return -10;
+  }
+  if (ad != NULL && ad_len > 0 &&
+      EVP_DecryptUpdate(ctx->ocb_ctx, NULL, &len, ad, ad_len) != 1) {
```

Comment:
(IIRC, it'll crash if you don't pass in `&len`, but what it writes to `len` is totally useless, so ignoring it here is correct. Ditto with encrypt.)